dronecaster.lua

```diff
@@ -57,6 +66,71 @@ done_init = false
 
 -- init & core
 --------------------------------------------------------------------------------
+
+-- midi
+function build_midi_device_list()
+  midi_devices = {}
+  for i = 1,#midi.vports do
+    local long_name = midi.vports[i].name
+    local short_name = string.len(long_name) > 15 and util.acronym(long_name) or long_name
+    table.insert(midi_devices,i..": "..short_name)
+  end
+end
+
+function midi_event(data)
+  -- global midi setting
+  if midi_enabled == 1 then
+    msg = midi.to_msg(data)
+    -- filter channel
+    if msg.ch == midi_channel then
+      if msg.type ~= "clock" then
+        tab.print(msg)
+      end
+      --note message
+      if msg.type == "note_on" then
+        params:set("note",msg.note)
+        -- amp control
+        print(midi_amp_control)
+        if midi_amp_control == 2 then
+          amp = math.min(msg.vel/100, 1)
+        end
+        if midi_amp_control == 3 then
+          amp = math.min(msg.key_pressure/100, 1)
+        end
+        if midi_amp_control ~= 0 then
+          params:set("amp",amp)
+        end
+      end
+      --cc message
+      if msg.type == "cc" then
+        if msg.cc == midi_amp_cc then
+          params:set("amp", (msg.val/127))
+        end
+        if msg.cc == midi_drone_cc then
+          minval = math.min(msg.val, #drones)
+          params:set("drone", minval)
+        end
+      end
+      -- program change message
+      if msg.type == program_change then
+        minval = math.min(msg.val, #drones)
+        params:set("drone", minval)
+      end
+    end
+    -- transport
+    if msg.type == "start" then
```

Comment:
Does this mean notes & transport are a package deal? Also if I sequenced the drones externally and then stopped the sequencer, it will silence the drone too.

Reply:
yes. Good point. I'll add an option to ignore transports ?

Reply:
Well, I could see an arguments for both ignoring transports and ignoring notes. I'm not even saying this is a bad thing, especially since selecting both would add UI/implementation complexity. I'd almost rather keep the whole shebang simpler and stick to a more opinionated set of behaviors. What do you think?

Reply:
It's fairly easy to add,  More complex behavior can be an option param. Transport behaviour can be "all", "ignore stop" or "ignore all".
Also not sure why you'd want to ignore notes; it's be strange to send that data but then want it ignored. Just filter by channel.

Reply:
> Also not sure why you'd want to ignore notes; it's be strange to send that data but then want it ignored. Just filter by channel.

That's a decent solution. I've set hardware to weird MIDI channels to avoid triggering it before. There have been a lot of surprising use cases for dronecaster, though, so I'm not ruling any "strange" stuff out!



> Also just to clarify: CC control over Hz is very unstable (i guess this is what you are thinking of). 127 steps for control for 0-20000hz is just not working out. It actually crashes easily. Note control is much better.

Agreed, that sounds like a mess!